orchagent/intfsorch.cpp

```diff
@@ -398,3 +409,83 @@ void IntfsOrch::removeIp2MeRoute(const IpPrefix &ip_prefix)
 
     SWSS_LOG_NOTICE("Remove packet action trap route ip:%s", ip_prefix.getIp().to_string().c_str());
 }
+
+/*
+ * Function runs basic sanity-checks to verify that the new interface's ip
+ * address being configured does not overlap with any of the existing interfaces.
+ *
+ * There are two possible scenarios in which we can encounter overlapping
+ * interfaces:
+ *
+ * 1) Obviously, a configuration error caused by a user creating overlapping
+ *    interfaces.
+ *
+ * 2) A configuration change made through 'ifconfig' tool. In this case, the
+ *    following set of netlink messages is always generated. The exception to
+ *    this message sequence is the case in which the new ip-addr contains a /8
+ *    netmask -- in that case only 2.1) and 2.2) are expected. In all other
+ *    cases, this is the expected behavior:
```

Comment:
I am a little confused, because this netlink message sequencing is what previous code handled.

Reply:
Right, the previous code (127-144) addresses 'ifconfig' special msg sequencing, and by doing that it verifies if there's a subnet that overlaps with the _same_ interface that is being configured. However, the goal there is not to detect overlaps, but to ensure that the route-deletion is consumed before the final route-addition. It happens to use overlap-detection logic coz it serves the purpose of identifying the 'ifconfig' case, but that's not the goal there.

On the other, my new code truly takes care of preventing overlaps with the rest of the system interfaces. Now, due to the unfortunate 'ifconfig' behavior, i have to address that special case too. See an example below:

Eth1: 10.1.1.0/24
Eth2: 11.1.1.0/24

Now you do: ifconfig Eth2 10.2.2.0 netmask 255.255.255.0

There's no real overlap above, but 'ifconfig' behavior demands more thinking from our side. With today's code, upon arrival of netlink route-add msg for 10/8, we will go ahead and push that route down to asic. With this fix's code, we will see the overlap between 10/8 and 10.1.1.0/24, and we will skip this netlink message. Eventually, the proper route-add msg for 10.2.2.0/24 will arrive, and only then we will push it down to asic.

So as you can see, my code has to deal with this 'ifconfig' behavior too, that's why i went ahead and wrote a full description of that issue, and tested all possible combinations i could think of.